pkg/provider/azure_vmss.go

```diff
@@ -1082,7 +1066,7 @@ func getConfigForScaleSetByIPFamily(config *compute.VirtualMachineScaleSetNetwor
 		}
 	}
 
-	return nil, fmt.Errorf("failed to find a IPconfiguration(IPv6=%v) for the scale set VM %q", IPv6, nodeName)
+	return nil, fmt.Errorf("failed to find an IPconfiguration (IPv6=%v) for the scale set VM %q", IPv6, nodeName)
```

Comment:
I think so and here is the doc: https://learn.microsoft.com/en-us/azure/virtual-network/ip-services/virtual-network-network-interface-addresses?tabs=nic-address-portal#ip-configurations

"A primary IP configuration:
 Has a [private](https://learn.microsoft.com/en-us/azure/virtual-network/ip-services/virtual-network-network-interface-addresses?tabs=nic-address-portal#private) [IPv4](https://learn.microsoft.com/en-us/azure/virtual-network/ip-services/virtual-network-network-interface-addresses?tabs=nic-address-portal#ipv4) address assigned to it. You can't assign a private [IPv6](https://learn.microsoft.com/en-us/azure/virtual-network/ip-services/virtual-network-network-interface-addresses?tabs=nic-address-portal#ipv6) address to a primary IP configuration"

"A secondary IP configuration:
Must have a private IPv4 or IPv6 address assigned to it. If the address is IPv6, the network interface can only have one secondary IP configuration."

Reply:
For ipv4, if we always return the first ipv4 IP config, can we make sure that this one is primary all the time?
For ipv6, how do we make sure the wanted ipv6 config is picked since there could be multiple ipv6 configs?

Reply:
> For ipv4, if we always return the first ipv4 IP config, can we make sure that this one is primary all the time?

When there's only 1 ip config, we pick the first. Otherwise, we check if it is the Primary one.
https://github.com/kubernetes-sigs/cloud-provider-azure/blob/589f009805948197db91bc225b356bd05795d094/pkg/provider/azure_vmss.go#L1023-L1032

> For ipv6, how do we make sure the wanted ipv6 config is picked since there could be multiple ipv6 configs?

If I understand the doc clearly, there won't be multiple ipv6 config. If there's an ipv6 config, then it must be the secondary ip config. As the doc says, "If the address is IPv6, the network interface can only have one secondary IP configuration". Could you please share a scenario with multiple ipv6 ip configs?

Reply:
Yeah that's for getPrimaryNetworkInterfaceConfiguration but in getConfigForScaleSetByIPFamily only the first ipv4 config is picked.

Reply:
Oh, now I get your point.
Yes, but currently all usage of `getConfigForScaleSetByIPFamily` is for IPv6 (with `if isIPv6`). Do you suggest to making `getConfigForScaleSetByIPFamily` IPv6 only?

Reply:
I suggest merging these two functions.

Reply:
Good idea. Will try. Thank you!

Reply:
@nilo19 PTAL